Integrate AsyncLM models from genlm-backend

benlebrun opened the pull request:
This PR updates the `CachedCausalLM` class to use the `AsyncLM` models from `genlm-backend`. The key changes are:

# Major Changes
* `CachedCausalLM` now wraps an `AsyncLM` instance instead of HuggingFace's `CausalLM`
* Added support for multiple backends through a `backend` parameter in `.from_pretrained`:
     - 'vllm': Uses `AsyncVirtualLM` (recommended for GPU)
     - 'hf': Uses `AsyncTransformer` (recommended for CPU)
     - 'mock': Uses `MockAsyncLM` (for testing/development)
    
    (all three are subclasses of `AsyncLM`)

* Moved caching and batching logic to `genlm-backend`'s `AsyncTransformer` class
* Vocabulary handling is now managed by `genlm-backend`:
    - Renamed `.vocab` to `.str_vocab` (with deprecation warning for `.vocab`)
    - Added new `.byte_vocab` attribute
    
* Added a `test` directory that tests whether the haiku and hard constraints examples run with each backend.
    
 * TODO: update documentation

# Additional Changes
* Added `run_example` functions to haiku and hard constraint examples for testing/benchmarking. I had to remove a few global references to the LLM, so the underlying `Model` classes were slightly modified (their functionality is the same though).
* Updated version number to reflect significant changes
* Replaced all `.vocab` references I came across with `.str_vocab`

# Compatibility
I tried to maintain backward compatibility where possible and added deprecation warnings for legacy features. 

One pain point is the `auth_token` parameter that is accepted by the `.from_pretrained` method. vLLM does not support authentication in this way, so we throw an error if `auth_token` is passed in with `backend='vllm'`. I also added a warning saying that the `auth_token` parameter will be removed in a future version, since I find it ugly to support this form of authentication for only one backend (and authenticating through the HuggingFace CLI is actually lower friction in the long-run, and avoids a bunch of `os.ENVIRON[...]` code).

gabegrand commented:
Hey @benlebrun, just wanted to say thank you so much for this contribution -- this looks *awesome* and I'm super excited to try it out.

Do you have any benchmarks on the speedup we see from the vLLM backend? I would be curious to see some performance comparisons, even just for the haiku / hard constraints examples.

gabegrand commented on pyproject.toml:
```diff
@@ -14,14 +14,12 @@ packages = [{include = "hfppl"}]
 nltk = {version = "^3.8.1", optional = true}
 pre-commit = "^3.7.1"
+genlm-backend = { git = "https://github.com/probcomp/genlm-backend.git" }
+ipykernel = "^6.29.5"
```
minor nit: is ipykernel actually needed or is this a dev dependency?

benlebrun replied:
yeah thanks for catching that! removed it

gabegrand commented on hfppl/distributions/lmcontext.py:
```diff
@@ -27,6 +27,7 @@ async def log_prob(self, x):
 
     async def sample(self):
         probs = np.exp(self.ctx.next_token_logprobs)
+        probs /= np.sum(probs)
```
I think this line should have been added already in #12 -- can you rebase on main?

benlebrun replied:
I rebased though this change still appears

benlebrun replied:
ok never mind, I just merged main into this branch to resolve this conflict. All good!

benlebrun commented:
@gabegrand I added a benchmark script to compare hf and vllm backends. here's the output I get by running that benchmark on an 80GB A100:

<img width="1219" alt="Screenshot 2025-01-22 at 6 52 39 PM" src="https://github.com/user-attachments/assets/50e7f435-225d-4e8d-aa49-fdb80297e7af" />

gabegrand reviewed (approved):
I've now tested this PR thoroughly and can attest to the following:
- I was able to reproduce Ben's 2-5x speedups with the vLLM backend (see below)
- The HuggingFace backend is backwards compatible with hfppl v0.1.0.
  - Existing hfppl code should behave normally, with the caveat that for HF/vLLM and GPU compatibility reasons, models are no longer loaded with 8-bit quantization by default. (see discussion in https://github.com/probcomp/hfppl/pull/26 and https://github.com/probcomp/genlm-backend/pull/1)
- Existing research code based on `hfppl` can be easily ported over to support the vLLM backend. 
  - The only change necessary (besides passing `backend=vllm`) is to use the async `.create()` factory pattern when instantiating `LMContext` objects. This does also require `Model.create()` to be used in place of `Model.__init__()`. The examples have been migrated over to reflect this new pattern.

Thank you @benlebrun for this outstanding piece of engineering work and for helping to resolve minor snags. Pending release of https://github.com/probcomp/genlm-backend, I think `hfppl v0.2.0` is good to ship!

<img width="1855" alt="Screenshot 2025-01-23 at 6 14 03 PM" src="https://github.com/user-attachments/assets/eb194bde-6e04-4eae-b55c-96f3e0834cb7" />

gabegrand commented:
Tagging @alex-lew for final review!

alex-lew commented:
@benlebrun and @gabegrand -- thanks for this fantastic work! Excited to take a closer look and start using. But before we merge: can we make genlm-backend a public repo? hfppl can't have a private repo as a dependency.

benlebrun commented:
@alex-lew I made `genlm-backend` public. Note that it has been moved to the `chi-collective` organization; dependency has been updated (and reverted to back an HTTPS link). l also explicitly specified `backend='hf'` in the README example to avoid potential issues with LMContext creation. We could also update this example to use the create methods, but that may be orthogonal to its purpose.

alex-lew commented:
@benlebrun Thanks! It looks like the link is still git@github.com and not https?

alex-lew reviewed (approved):
Changed the link to HTTPS, and I think it looks good now! Thanks again for this -- really awesome to have the new backend working!

benlebrun commented:
@alex-lew @gabegrand I may have identified a way to properly handle `next_token_logprobs_sync` for the vllm backend in `genlm_backend`, which may mean that we no longer need the ugly `create` methods. Stay tuned for more; flagging now because we might want to hold off merging if it turns out we actually don't need that API change.

benlebrun commented:
@gabegrand @alex-lew sorry for delay on this. I am pretty confident that `next_token_logprobs_sync` for the vllm backend will no longer fail in asynchronous contexts. Note though that I was not able to reproduce the setting which lead to the original issue @gabegrand was having with the old implementation. 

A few changes as a result of this:

* I removed the use of the async `.create` method from the examples; `__init__` should work in synchronous and asynchronous contexts for all backends. Note that I kept the `.create` method in the `LMContext` class since it may be useful for users who want to take advantage of the async batching logic when creating `LMContext`s concurrently. 
* I added a test for LMContext initialization using `__init__` and `.create` in synchronous and asynchronous contexts.

gabegrand reviewed (approved):
Hi @benlebrun, thanks for following up! I just did a quick test and confirmed that after your changes to `next_token_logprobs_sync`, creating `LMContext` objects inside sync environments no longer results in hanging. Thanks so much for going the extra mile to fix this.

FWIW, I actually kind of like how the `.create()` syntax works in a setting where you have many hierarchical subclasses of some base `Model` class. Because the subclass constructors don't have to pass through a bunch of parameters to the `LMContext` constructor, it makes the `__init__()` methods cleaner. So I don't see any harm in keeping `.create()` in the API for now :)

One final thing that occurred to me (unrelated to the above). I had configured `vllm` to be an optional dependency for `hfppl` (so you would do `poetry install --with vllm`). However, we currently have `CachedCausalLM` configured to use `vllm` by default when CUDA is available, regardless of whether the user has `vllm` installed, which could lead to some confusion. I would like to have things work out-of-box as much as possible. 

In that spirit, I think we should either:
- Make `vllm` a default dependency, OR
- Add some logic to make `CachedCausalLM` aware of whether `vllm` is available as a backend and act appropriately

@benlebrun what do you think?

Aside from that, I would definitely move to have this PR merged ASAP so that we can base future development on it.

benlebrun commented:
@gabegrand I added a check in `llms.py` for vllm availability and made it so that the vllm backend is only used if cuda is available and if vllm is installed. 

I am good on merging, so feel free to pull trigger when you see fit!
